### PR TITLE
LUCENE-10200: Correct outdated instruction in the demo tutorial

### PR DESCRIPTION
**Pull request by mocobeta**
### Description (or a Jira issue link if you have one)

This is a minor update for `demo` module documentation.
I had a chance to run the demo app and noticed that commands in the tutorial use `CLASSPATH` (and some description has been outdated); maybe it'd be worth updating to use modulepath instead of classpath?

The latest tutorial:
Screenshot from 2022-05-24 19-28-13

Updated tutorial in this patch:
Screenshot from 2022-05-25 22-43-32

### CODE REVIEW

**Comment by mocobeta** [comment]
note: `modules-thirdparty` should be included on the module path since it depends on `hppc` via `lucene-facet`. Also it seems good to add `--add-modules jdk.unsupported` when running `SearchFiles`, otherwise, it emits (a little intimidating) "WARNING: Unmapping is not supported, ..." message.

**Comment by msokolov** [comment]
I found a few typos; if you have a moment maybe you could fix while you're updating?

I'm also curious - is it still possible to use the old way (with CLASSPATH / -cp)? Also, is MODULEPATH an environment variable like CLASSPATH? If it is, do we also need the command-line arguments as shown?  I guess the previous documentation didn't really explain how to set the CLASSPATH, but given that MODULEPATH is a new thing and many users may be unfamiliar with it, maybe we should take the chance to educate a bit how to use it here.

**Comment by mocobeta** [comment]
>  Also, is MODULEPATH an environment variable like CLASSPATH? 

Thanks, it's a very good point; to the best of my knowledge, there is no environment variable to implicitly set module paths such as `MODULEPATH`. The fully capitalized term is misleading, I updated it to the normal "module path".

> I'm also curious - is it still possible to use the old way (with CLASSPATH / -cp)?

Yes, classpath will continue to remain and I think it's unlikely to happen to be dropped.

> I found a few typos; if you have a moment maybe you could fix while you're updating?

Sure, could you tell me the lines we should fix?

**Comment by mocobeta** [comment]
> I guess the previous documentation didn't really explain how to set the CLASSPATH, but given that MODULEPATH is a new thing and many users may be unfamiliar with it, maybe we should take the chance to educate a bit how to use it here.

Instead of explaining what is java module system and its usage, I would give concrete commands that work by copy-pasting for this tutorial.

**Comment by msokolov** [comment]
>>  I found a few typos; if you have a moment maybe you could fix while you're updating?

> Sure, could you tell me the lines we should fix?

Hmm I added comments above, but github didn't post them yet - coming soon...

**Review by msokolov — COMMENTED** [review]
ok the comments were part of a "review" so they didn't get posted with my other comments...

File: lucene/demo/src/java/overview.html [hunk 1]
```diff
@@ -49,36 +49,35 @@ <h2 class="boxed">About the Demo</h2>
...
 </pre>
 This will produce a subdirectory called <span class="codefrag">index</span>
 which will contain an index of all of the Lucene source code.
 <p>To <b>search the index</b> type:</p>
 <pre>
-    java org.apache.lucene.demo.SearchFiles
+    java --module-path modules:modules-thirdparty --add-modules jdk.unsupported --module org.apache.lucene.demo/org.apache.lucene.demo.SearchFiles
 </pre>
 You'll be prompted for a query. Type in a gibberish or made up word (for example: 
 "superca<!-- need to break up word in a way that is not visibile so it doesn't cause this ile to match a search on this word -->lifragilisticexpialidocious").
```
[the remark this answers is not in the document]
Comment (mocobeta) [reply lucene/demo/src/java/overview.html]: Fixed in https://github.com/apache/lucene/pull/923/commits/d400d7bed703999c050407f9f5b6cf9c0f66b748 - I still can't detect typos in English by just quickly skimming :)

**Comment by mocobeta** [comment]
I hooked this on LUCENE-10200 - actually, the tutorial has been obsoleted by the change in the way to assemble the binary distribution.
Still, we can stick to classpath though, I feel like it'd be clearer to switch to module path to align with the binary release structure (as well as the Luke launch script).

**Comment by msokolov** [comment]
Thanks for fixing those (pre-existing) typos. On the classpath/module path change I have mixed feelings. On the one hand, we should broadcast that we are now fully modularized and support using module paths to declare dependencies on code. On the other hand, I don't even know how to "put jars on my module path" yet. Maybe I'm just a stick-in-the-mud, but modules still seems very new and I suspect many (most) Java devs probably haven't yet figured it out and are still using class-path? So, I'm not sure what that means for this documentation. It doesn't seem as if this is the right place to explain modules and class paths, but we want to make this accessible and easy to use. Maybe we could include both sets of directions?

**Comment by dweiss** [comment]
Modular are not new, they're just not widespread... I agree with @mocobeta that if you provide an explicit command line then there is little harm in not explaining all the options. The problem with classpath is that you need to include all JARs individually  this is handled much better by modules since you include the directory path (not each individual module JAR).

This said, I don't have a strong opinion about going either way.

**Comment by mocobeta** [comment]
Anyway the outdated instruction would need to be corrected.
Maybe we can write both working commands for classpath and module path? Let me adjust it...

**Comment by msokolov** [comment]
>  The problem with classpath is that you need to include all JARs individually this is handled much better by modules since you include the directory path (not each individual module JAR).

Oh, that is better! Although I have gotten used to using `*.jar` in classpaths which helps shrink them down to be more manageable

**Comment by mocobeta** [comment]
I updated the text so that we have both working commands for module path and classpath in it. Please see the updated screenshot in the PR description to see how it looks, thanks.

File: lucene/demo/src/java/overview.html [hunk 2]
```diff
@@ -49,40 +49,49 @@ <h2 class="boxed">About the Demo</h2>
...
-index!</b> Assuming you've set your CLASSPATH correctly, just type:</p>
+index!</b> Just type either command:</p>
+<p>With module path</p>
 <pre>
-    java org.apache.lucene.demo.IndexFiles -docs {path-to-lucene}
+    java --module-path modules:modules-thirdparty --module org.apache.lucene.demo/org.apache.lucene.demo.IndexFiles -docs {path-to-lucene}
+</pre>
+<p>With classpath</p>
+<pre>
+    java -cp "modules/*:modules-thirdparty/*" org.apache.lucene.demo.IndexFiles -docs {path-to-lucene}
```
[the remark this answers is not in the document]
Comment (dweiss) [reply lucene/demo/src/java/overview.html:79]: I don't think it'll work. Java doesn't expand wildcards in arguments. Also, the path delimiter varies between platforms (Windows uses a semicolon)...

[the remark this answers is not in the document]
Comment (mocobeta) [reply lucene/demo/src/java/overview.html:79]: It works, I confirmed this command. There is "Understanding class path wildcards" section in the documentation.
https://docs.oracle.com/javase/7/docs/technotes/tools/windows/classpath.html

As for the delimiter, I didn't think we should list up all commands for Windows and Linux/Mac; then I committed Windows here... sorry but the tutorial is written for Unix-like platform from the beginning.

**Comment by mocobeta** [comment]
I first thought it is sufficient to have a concrete working command for a Unix-like platform that based on module path, seems like things are not so obvious. I don't know what should I do here - I'll leave it for now.

**Comment by mocobeta** [comment]
I was just unsure where to go for improving it. There will be a better way that does not overwhelm the readers of this tutorial with unfalimiar concepts or too wordy descriptions, I haven't found it yet though. Maybe, I'll open another one sometime.

**Comment by dweiss** [comment]
Umm.. I agree with Mike, Tomoko - the improvement was there and evident. Can we commit it in and maybe reiterate if we find a better wording in the future? What you did was way better than it had been before.

**Comment by mocobeta** [comment]
I'm unsure if the change improves it. Thinking calmly there are two questions in the change here.
1. The start commands for the demo app are no longer platform agnostic.
    - The old one works on any platform since it does not include any file paths in it - it picks CLASSPATH environment variable.
2. Put all jars in a directory into classpath with `*` is not great.
    - Old one says "Put all four of these files in your Java CLASSPATH." - yes it's the correct way to encourage people to find necessary jars and construct classpath on their own, except for the list of jars is obsoleted.

As we saw things seem not so obvious as I first thought, I would postpone it for now; anyway it does not a big deal I think.

**Comment by mocobeta** [comment]
Also thank you both for your comments - at least I learned the subtle difficulties related to launching commands (again).
